Objects/setobject.c

```diff
@@ -118,9 +120,15 @@ set_add_entry(PySetObject *so, PyObject *key, Py_hash_t hash)
   restart:
 
     mask = so->mask;
-    i = (size_t)hash & mask;
     freeslot = NULL;
-    perturb = hash;
+
+    if (_PyReproducibleFrozenSet_CheckExact(so)) {
+        i = 0;
+        perturb = 0;
+    } else {
+        i = (size_t)hash & mask;
+        perturb = hash;
+    }
```

Comment:
Unfortunately, this impacts the performance of normal set operations, and given that is a basic type of the interpreter I am not very convinced about it. This is one of the reasons I said:

> and a new type, which I am not very fond of) 

Because a new type will require a new set of independent functions on the required path to not impact the regular set and frozenset and that is much more code than what I am confortable with.

On the other hand, others may have other views on this, so I am happy to be convinced otherwise :)